rbd/diff_iterate.go

```diff
@@ -0,0 +1,133 @@
+			uint64_t ofs, uint64_t len,
+			uint8_t include_parent, uint8_t whole_object,
+			void *cb,
+			void *arg) {
+	return rbd_diff_iterate2(image, fromsnapname, ofs, len, include_parent, whole_object, cb, arg);
+}
+*/
+import "C"
+
+import (
+	"unsafe"
+)
+
+// DiffIncludeParent values control if the difference should include the parent
+// image.
+type DiffIncludeParent uint8
+
+// DiffWholeObject values control if the diff extents should cover the whole
+// object.
+type DiffWholeObject uint8
+
+// DiffIterateCallback defines the function signature needed for the
+// DiffIterate callback.
+//
+// The function will be called with the arguments: offset, length, exists, and
+// config. The offset and length correspond to the changed region of the image.
+// The exists value is set to zero if the region is known to be zeros,
+// otherwise it is set to 1. The config value is the DiffIterateConfig that was
+// passed to DiffIterate and is meant to be used for reference back to the
+// original caller as well as a way to pass arbitrary data to the callback, via
+// the Data parameter.
+//
+// The callback can trigger the iteration to terminate early by returning
+// a non-zero error code.
+type DiffIterateCallback func(uint64, uint64, int, DiffIterateConfig) int
+
+// DiffIterateConfig is used to define the parameters of a DiffIterate call.
+// Callback, Offset, and Length should always be specified when passed to
+// DiffIterate. The other values are optional.
+type DiffIterateConfig struct {
+	SnapName      string
+	Offset        uint64
+	Length        uint64
+	IncludeParent DiffIncludeParent
+	WholeObject   DiffWholeObject
+	Callback      DiffIterateCallback
+	Data          interface{}
+}
+
+const (
+	// ExcludeParent will exclude the parent from the diff.
+	ExcludeParent = DiffIncludeParent(0)
+	// IncludeParent will include the parent in the diff.
+	IncludeParent = DiffIncludeParent(1)
+
+	// DisableWholeObject will not use the whole object in the diff.
+	DisableWholeObject = DiffWholeObject(0)
+	// EnableWholeObject will use the whole object in the diff.
+	EnableWholeObject = DiffWholeObject(1)
+)
+
+// DiffIterate calls a callback on changed extents of an image.
+//
+// Calling DiffIterate will cause the callback specified in the
+// DiffIterateConfig to be called as many times as there are changed
+// regions in the image (controlled by the parameters as passed to librbd).
+//
+// See the documentation of DiffIterateCallback for a description of the
+// arguments to the callback and the return behavior.
+//
+// Implements:
+//  int rbd_diff_iterate2(rbd_image_t image,
+//                        const char *fromsnapname,
+//                        uint64_t ofs, uint64_t len,
+//                        uint8_t include_parent, uint8_t whole_object,
+//                        int (*cb)(uint64_t, size_t, int, void *),
+//                        void *arg);
+func (image *Image) DiffIterate(config DiffIterateConfig) error {
```

Comment:
I think I prefer a DiffIterateCallback struct that contains the cb+arg only. The other arguments for rbd_diff_iterate2() can be passed to DiffIterate as they are. Placing everything in a DiffIterateConfig seems too far off the 'stick to the librbd API' for me.

Reply:
> I think I prefer a DiffIterateCallback struct that contains the cb+arg only. The other arguments for rbd_diff_iterate2() can be passed to DiffIterate as they are. Placing everything in a DiffIterateConfig seems too far off the 'stick to the librbd API' for me.

Let me attempt to justify why this isn't much of a deviation from the librbd api, and that's it is worth it:
The current DiffIterateConfig contains struct fields that are one-to-one mappings of the args of the C api call. The struct fields are in the same order as the C call. The C api call has eight arguments which is above (my) ability to remember what order they go in. By moving the whole argument set into a dedicated struct we make it more self-documenting and more difficult to accidentally misuse. The only parameters where the default values for their type don't work well are Length and Callback. Leaving some of the fields out to fall back to the zero-values increases the convenience without losing any of the power of the full api.

I think that if we decide to only do arguments then there's no point in doing a struct that's  public. To me that would be more deviating mixing positional args with new type.

I'd argue that I'd rather keep the DiffIterateConfig but only pass Data to the callback rather than the config. That's the part I thought would be more controversial. :-)

Reply:
So, I have gone through this an other time. The link to [Passing callbacks and pointers to Cgo](https://eli.thegreenplace.net/2019/passing-callbacks-and-pointers-to-cgo/) explains a little more why I had difficulties passing an address to a Go object to the C shim callback for the watchers (other PR). [cgo/Function pointer callbacks](https://github.com/golang/go/wiki/cgo#function-pointer-callbacks) mentions it, sortfof, but not as clear.

I am not very happy with the `cmap`, but now I understand the need for it. So, ok, this works, lets stick to this approach.

Passing DiffIterateConfig is indeed something that I would not prefer. But I also do not object too much to it. It is only ugly to see when other callbacks do not need a DiffIterateConfig like `rbd_update_watch()`. In that case, the callback would get called with `data interface{}` directly.

So, for overall similarity between callback features, I would prefer to have the callback get called with `Data` and not a Config-object.

Reply:
Are you (mildly) unhappy with the need for the cmap or the implementation? I can't do much about the former. :-)
I will change the callback to accept a data interface{}.

Reply:
> Are you (mildly) unhappy with the need for the cmap or the implementation?

I don't like storing callback items in a map, and fetching them by integer. But, well, this is a limitation of CGO, so nothing we can do about it. The implementation itself seems fine, I understand how it works and could not find any problems while reviewing it (several times now).